### ISSUE TITLE
Don't make IRequestHandler<TRequest, TResponse> variant

### ISSUE THREAD

**Issue by dotnetjunkie** [christophdebaene/CleanArchitecture#3]
Currently, the `IRequestHandler` abstraction is defined as variant (with `in` and `out` keywords):

``` c#
interface IRequestHandler<in TRequest, out TResponse> { }
```

The use of `in` and `out` keywords implies that multiple implementations are applicable to one single request message, but this usually doesn't make sense for request handlers. For request handlers, there is always a one-to-one-to-one mapping (from request to response with one handler implementation). If there is a `ShipOrder` request object with a `ShipOrderRequestHandler` implementation, it wouldn't make much sense to have a `ShipSpecialOrder` type that derives from `ShipOrder` that can be handled by the same `ShipOrderRequestHandler`.

There are places where variant interfaces make a lot of sense. `IValidator<in T>` for instance and `IEventHandler<in TEvent>`. These are abstractions that are usually registered as collection, meaning that there can be multiple validators for a single object or multiple handlers that can process a particular event. For instance, when a `SpecialOrderShipped : OrderShipped` event is published, it would be fine when it is picked up by an `IEventHandler<OrderShipped>` implementation.

In other words, the use of `in` and `out` keywords in the abstraction is confusing to anyone who reads or maintains it when used when it will not be used. I often see developers use variant interfaces because Resharper suggests this. Don't listen to Resharper! It doesn't know anything about your application design; it only looks at the particular interface.

Same holds for the `IRequest<out TResponse>` interface.

**Comment by christophdebaene** [comment]
Indeed, subtle but important!
Thx for the tip, I will adapt the code.
